### DISCUSSION TITLE
How can I configure OpenURI?

### DISCUSSION THREAD

**Discussion by amano-kenji** [flatpak/xdg-desktop-portal#1348] [Questions]
I'm using xdg-desktop-portal-gtk on sway, and I don't know how to configure OpenURI. I want to associate http links with a specific web browser.

**Comment by smcv** [comment]
It follows the MIME Apps Specification: https://specifications.freedesktop.org/mime-apps-spec/latest/

https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/using_the_desktop_environment_in_rhel_8/configuring-file-associations_using-the-desktop-environment-in-rhel-8#overriding-default-registered-application-for-individual-users_configuring-file-associations is a more tutorial-style description (written for RHEL 8, but it should be equally valid for any fd.o-based distribution).

**Reply by amano-kenji** [reply]
But, dino XMPP client ignores ~/.config/mimeapps.list. It just opens brave browser although ~/.config/mimeapps.list designates another browser for opening http links.

**Reply by amano-kenji** [reply]
According to dbus-monitor,
```
method call time=1713954708.381215 sender=:1.11 -> destination=:1.15 serial=2264 path=/org/freedesktop/impl/portal/PermissionStore; interface=org.freedesktop.impl.portal.PermissionStore; member=Lookup
   string "desktop-used-apps"
   string "x-scheme-handler/https"
error time=1713954708.381495 sender=:1.15 -> destination=:1.11 error_name=org.freedesktop.portal.Error.NotFound reply_serial=2264
   string "No entry for x-scheme-handler/https"
signal time=1713954708.384354 sender=:1.11 -> destination=(null destination) serial=2265 path=/org/gtk/gio/DesktopAppInfo; interface=org.gtk.gio.DesktopAppInfo; member=Launched
   array of bytes "/usr/share/applications/brave-browser.desktop" + \0
```

It fails to find an entry for x-scheme-handler/https, but ~/.config/mimeapps.list has an entry for that.

**Reply by amano-kenji** [reply]
I guess it is a bug in xdg-desktop-portal-gtk?

**Reply by smcv** [reply]
The portal should usually be doing the same thing as `gio open https://whatever` and `gio mime x-scheme-handler/https`, whatever that is.

**Reply by amano-kenji** [reply]
`gio open https://duckduckgo.com` and `gio mime x-scheme-handler/https` behave as expected. They respect ~/.config/mimeapps.list. However, when I click a link on dino XMPP client, brave browser is opened. Brave browser is the first recommended application in the absence of any default application for x-scheme-handler/https.

dino seems to invoke OpenURI.

**Reply by Rosuavio** [reply]
I have the same issue. Brave always open (when opening a url from all my apps) and I believe I have everything configured correctly. 

```cmd
❯ xdg-settings get default-url-scheme-handler http
firefox.desktop

❯ cat .local/share/applications/mimeapps.list 
[Added Associations]
x-scheme-handler/about=firefox.desktop
x-scheme-handler/http=firefox.desktop
x-scheme-handler/https=firefox.desktop
x-scheme-handler/unknown=firefox.desktop

[Default Applications]
x-scheme-handler/about=firefox.desktop
x-scheme-handler/bitwarden=Bitwarden.desktop
x-scheme-handler/http=firefox.desktop
x-scheme-handler/https=firefox.desktop
x-scheme-handler/unknown=firefox.desktop

[Removed Associations]

❯ echo $NIXOS_XDG_OPEN_USE_PORTAL 
1
```

`gio open https://duckduckgo.com` opens firefox, but `xdg-open https://duckduckgo.com` opens brave.

**Reply by Rosuavio** [reply]
I know that `xdg-open` calls `OpenURI` I am assuming that on my system it is handled by the `OpenURI` handler here. But I am trying to dig into where `xdp_dbus_open_uri_complete_open_uri` leads and I can't find it anywhere on the internet.

**Reply by q962** [reply]
This is the same issue as in my post.
https://github.com/flatpak/xdg-desktop-portal/discussions/1874

A temporary solution is:

``` 
rm ~/.local/share/flatpak/db/desktop-used-apps
pkill -f xdg-permission-store
```

**Reply by Rosuavio** [reply]
I don't have that file on my machine.

**Reply by amano-kenji** [reply]
I don't use flatpak.

**Reply by q962** [reply]
What is your portal frontend?

**Reply by q962** [reply]
The current project is my portal frontend.

Since the path I provided doesn't exist, there must be a problem somewhere. I can provide my troubleshooting steps:

OpenUri will first call the get_latest_choice_info function.\
And the filename "desktop-used-apps" is provided here.
https://github.com/flatpak/xdg-desktop-portal/blob/3dde1a47e474513d3a732720b03e9d624a178dee/src/open-uri.c#L138-L139

Then the desktop-used-apps file will be read in the xdg-permission-store program.
https://github.com/flatpak/xdg-desktop-portal/blob/3dde1a47e474513d3a732720b03e9d624a178dee/document-portal/xdg-permission-store.c#L68-L72

**Reply by q962** [reply]
If you are using the xdg-open program, it might be a script. You can check the output using 'bash -x'.

**Reply by amano-kenji** [reply]
I think it's a portal backend. xdg-desktop-portal is the frontend. I use xdg-desktop-portal-gtk backend.

**Reply by q962** [reply]
Based on my observation, this issue is currently unrelated to the backend.

The backend only provides the AppChoose window, but you are not currently seeing a pop-up window.

Therefore, the best approach is to debug xdg-desktop-portal and xdg-permission-store.

I've listed the lines of code that need debugging. You can try it out. Please note that you'll need to download the debuginfo file.
